### DISCUSSION TITLE
Where are world files stored?

### DISCUSSION THREAD

**Discussion by nielsmittertreiner** [Pryaxis/TShock#2021] [Q&A]
I am hosting a small Terraria server with TShock for my friends and I, and I was wondering where the world files are being saved, I wanted to download them for other uses.

I've been looking around quite a lot but I can't seem to find it.

Thanks in advance!

**Comment by hakusaro** [comment]
It depends on the operating system. What operating system are you on?

**Reply by 3y3c0ugh** [reply]
I'm having this same issue, would you be able to assist?

I just setup a fresh install of Raspbian on my Pi4. Installed mono, downloaded tshock, made a new world, saved said world.

I've done this before on a Pi but with the old 1.3.5 version of Terraria. When I had it set up for that I could find saved worlds under ~/.local/share/Terraria/Worlds/

That directory doesn't exist now on my fresh install of OS and TShock. However, when I run Tshock it lists and loads my new world I created. Where is it being stored now though? I'd like to import old worlds I have backed up. Thanks!

**Reply by Cod3ingMast3r** [reply]
any luck?

**Reply by Lyxilion** [reply]
Same issue, looking for an anwser for hours now....

**Reply by 3y3c0ugh** [reply]
@Lyxilion @Cod3ingMast3r 
Yes, it was a pain in the ass to figure out but I did. Probably just as aggravated as you all got.
The old database files were in ~/.local/share/Terraria/Worlds/
They are now in /root/.local/share/Terraria/Worlds/
If you try and cd into /root/... you will get Access Denied. You can't sudo then change directory because sudo only works for executables, and cd is builtin shell.
Therefore, you have to issue a "sudo -s" to start a root level command shell. From there you can then "cd /root/.local/share/Terraria/Worlds/" once your done moving/adding files, simply issue an "exit" to return back to the original pi command shell.

**Reply by rautamiekka** [reply]
There's no way TShock could write into `root`'s folders unless TShock was running as one.

**Reply by 3y3c0ugh** [reply]
And that's correct. I can't get my server to run using just "mono TerrariaServer.exe" it throws a fatal startup exception that it could not write to "ServerLog.txt" because the path is being denied access. So I run all my instance as "sudo mono TerrariaServer.exe".

**Reply by rautamiekka** [reply]
^ What part of the full path ? If it just says the `ServerLog.txt`, then either
1) you got the file there already and it's write-protected
2) some setting blocks it, but frankly the latter makes little sense
3) your home folder, or path of the path under home, is write-protected.

See if the AppArmor or SELinux is somehow conffed to block something. I haven't dealt much with either (Ubuntu uses SELinux, and it hasn't given me much any problems during the years), but I'm pretty sure at least SELinux can be conffed to very selectively block file ops.

You can further debug the problem by this command:
```
strace --output=TerrariaServer.exe.log mono TerrariaServer.exe
```
It'll create a big log which you can open in an editor to see what it does. A good idea to start reading from the bottom since the errors should be there.

**Reply by 3y3c0ugh** [reply]
The whole path was `/home/pi/tshock/ServerLog.txt`. It more so sounds to me like 3 from your list. It seems like my home folder is write-protected. I think AppArmor was recently put in with the Buster distro for Pi's if I'm not mistaken, but I don't think it's enabled by default. I did however check in `/var/log/syslog` and `/var/log/kern.log` but don't see any errors or exceptions.

Probably not ideal, but I issued a `sudo chmod 777 /home/` and then went to tshock and ran `mono TerrariaServer.exe` which started and threw no exceptions. Now it doesn't see any of my worlds because it can put them in `/home/pi/.local/share/Terraria/Worlds/` now..

I feel stupid because I never bothered to check the actual permissions using `ls -l /home/` before I did it to see what they were set as. So I did the better thing and set it back to chmod 754. I then issued a `chown pi /home/` which I feel better about and also solved that problem. Now I need to move my worlds back from the root folder...

Lol I appreciate you helping me solve this, even though I wasn't too worried about how this ran. It's a Raspberry Pi 4 only used to run the TShock server and nothing else. Thanks again!
